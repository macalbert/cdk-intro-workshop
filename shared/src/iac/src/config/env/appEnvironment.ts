export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
