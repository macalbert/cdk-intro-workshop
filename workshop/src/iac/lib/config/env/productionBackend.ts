import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
