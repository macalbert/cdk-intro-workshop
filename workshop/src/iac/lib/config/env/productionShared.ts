import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
