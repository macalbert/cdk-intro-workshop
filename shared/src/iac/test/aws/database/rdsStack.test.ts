import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
    InstanceClass,
    InstanceSize,
    InstanceType,
    SubnetType,
    Vpc,
} from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";
import {
    RdsStack,
    type RdsStackProps,
} from "../../../src/aws/database/rdsStack";
import { AppEnvironment } from "../../../src/config/env/appEnvironment";

describe("RdsStackTest", () => {
    const LEGACY_VPC_CIDR = "10.0.0.0/16";
    test("Should_createRdsStack_When_ParamsAreValids", () => {
        // Arrange
        const app = new App();
        const stack = new Stack(app, "TestStack");

        const mockProps: RdsStackProps = {
            vpc: createVpc(stack),
            databaseName: "testdb",
            databaseUserName: "testuser",
            backupRetentionDays: 7,
            storageSizeGb: 20,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_15,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: "test-repo",
            envName: AppEnvironment.Production,
            stackName: "test-rds-stack",
            name: "TestRDS",
        };

        // Act
        const rdsStack = new RdsStack(stack, mockProps);

        // Assert
        const template = Template.fromStack(rdsStack);
        expect(template.toJSON()).toMatchSnapshot("RdsStackTest");

        template.hasResourceProperties("AWS::RDS::DBInstance", {
            DBInstanceIdentifier: mockProps.databaseName,
            DBName: mockProps.databaseName,
            Engine: "postgres",
            MasterUsername: mockProps.databaseUserName,
            DBInstanceClass: `db.${mockProps.instanceType.toString()}`,
            PreferredMaintenanceWindow: "Sun:03:00-Sun:04:00",
            EngineVersion:
                mockProps.engineVersion.postgresFullVersion.toString(),
        });

        template.hasResourceProperties("AWS::EC2::SecurityGroup", {
            GroupName: `${mockProps.githubRepo}Database`,
            SecurityGroupIngress: [
                {
                    FromPort: 5432,
                    ToPort: 5432,
                    IpProtocol: "tcp",
                    Description: "Allow PostgreSQL inbound from VPC",
                },
                {
                    FromPort: 5432,
                    ToPort: 5432,
                    IpProtocol: "tcp",
                    CidrIp: LEGACY_VPC_CIDR,
                    Description: "Allow PostgreSQL inbound from legacy VPC",
                },
            ],
        });

        template.resourceCountIs("AWS::SecretsManager::Secret", 1);

        template.hasResourceProperties("AWS::SecretsManager::Secret", {
            Name: "testdbPostgresqlCredentialsProduction",
        });

        const secretResources = template.findResources(
            "AWS::SecretsManager::Secret",
        );
        const secretResource = Object.values(secretResources)[0];
        expect(secretResource.Properties.Description).toContain("testdb");
        expect(secretResource.Properties.Description).toContain(
            "Database Credentials",
        );

        template.hasResourceProperties("AWS::SSM::Parameter", {
            Description: "PostgreSQL string connection TestRepo DB Production",
            Name: "/M47.TestRepo.Apps.Minimal.Api/Production/Database/ConnectionStrings/TestdbPostgres",
            Type: "String",
        });

        template.hasResourceProperties("AWS::SSM::Parameter", {
            Description: "PostgreSQL string connection TestRepo DB Production",
            Name: "/M47.TestRepo.Apps.WorkerService/Production/Database/ConnectionStrings/TestdbPostgres",
            Type: "String",
        });
    });

    function createVpc(scope: Stack) {
        return new Vpc(scope, "MockVpc", {
            maxAzs: 2,
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: "Public",
                    subnetType: SubnetType.PUBLIC,
                },
                {
                    cidrMask: 24,
                    name: "Private",
                    subnetType: SubnetType.PRIVATE_WITH_EGRESS,
                },
            ],
        });
    }
});
