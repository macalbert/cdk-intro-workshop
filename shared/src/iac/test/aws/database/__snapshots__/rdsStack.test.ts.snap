// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RdsStackTest Should_createRdsStack_When_ParamsAreValids: RdsStackTest 1`] = `
{
  "Outputs": {
    "SecretARN": {
      "Value": {
        "Ref": "postgresqlCredentials71085D2E",
      },
    },
    "SecretFullARN": {
      "Value": {
        "Ref": "postgresqlCredentials71085D2E",
      },
    },
    "SecretName": {
      "Value": {
        "Fn::Select": [
          0,
          {
            "Fn::Split": [
              "-",
              {
                "Fn::Select": [
                  6,
                  {
                    "Fn::Split": [
                      ":",
                      {
                        "Ref": "postgresqlCredentials71085D2E",
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    },
    "TestRDSProductionPostgresDbName": {
      "Export": {
        "Name": "test-repo-Production-PostgresDbName",
      },
      "Value": "testdb",
    },
    "TestRDSProductionPostgresEndpoint": {
      "Export": {
        "Name": "test-repo-Production-PostgresEndPoint",
      },
      "Value": {
        "Fn::GetAtt": [
          "PostgresDatabase0A8A7373",
          "Endpoint.Address",
        ],
      },
    },
    "TestRDSProductionPostgresUserName": {
      "Export": {
        "Name": "test-repo-Production-PostgresUserName",
      },
      "Value": "testuser",
    },
    "apiParameterPostgresConnectionParameterNameOutput": {
      "Description": "The name of the SSM parameter",
      "Export": {
        "Name": "apiParameterPostgresConnectionParameterName",
      },
      "Value": {
        "Ref": "apiParameterPostgresConnection9FF9728D",
      },
    },
    "workerParameterPostgresConnectionParameterNameOutput": {
      "Description": "The name of the SSM parameter",
      "Export": {
        "Name": "workerParameterPostgresConnectionParameterName",
      },
      "Value": {
        "Ref": "workerParameterPostgresConnection552F442F",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DatabaseSubnetGroup": {
      "Properties": {
        "DBSubnetGroupDescription": "test-rds-stacksubnet group",
        "DBSubnetGroupName": "test-rds-stacksubnet group",
        "SubnetIds": [
          {
            "Fn::ImportValue": "TestStack:ExportsOutputRefMockVpcPrivateSubnet1Subnet671F0665FE0D694A",
          },
          {
            "Fn::ImportValue": "TestStack:ExportsOutputRefMockVpcPrivateSubnet2Subnet9DD10F8AB112975D",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "ParameterGroup5E32DECB": {
      "Properties": {
        "Description": "Parameter group for postgres15",
        "Family": "postgres15",
        "Parameters": {},
      },
      "Type": "AWS::RDS::DBParameterGroup",
    },
    "PostgresDatabase0A8A7373": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AllocatedStorage": "20",
        "AllowMajorVersionUpgrade": false,
        "AutoMinorVersionUpgrade": true,
        "BackupRetentionPeriod": 7,
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.t3.micro",
        "DBInstanceIdentifier": "testdb",
        "DBName": "testdb",
        "DBParameterGroupName": {
          "Ref": "ParameterGroup5E32DECB",
        },
        "DBSubnetGroupName": {
          "Ref": "DatabaseSubnetGroup",
        },
        "DeletionProtection": true,
        "Engine": "postgres",
        "EngineVersion": "15",
        "MasterUserPassword": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:",
              {
                "Ref": "postgresqlCredentials71085D2E",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": "testuser",
        "MaxAllocatedStorage": 22,
        "MonitoringInterval": 60,
        "MonitoringRoleArn": {
          "Fn::GetAtt": [
            "PostgresDatabaseMonitoringRole8AB0DCC9",
            "Arn",
          ],
        },
        "PreferredMaintenanceWindow": "Sun:03:00-Sun:04:00",
        "PubliclyAccessible": false,
        "StorageEncrypted": true,
        "StorageType": "gp2",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Production",
          },
          {
            "Key": "Project",
            "Value": "test-repo",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-repo-testrds-production-stack",
          },
        ],
        "VPCSecurityGroups": [
          {
            "Fn::GetAtt": [
              "databaseSecurityGroupC7B2A60C",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Delete",
    },
    "PostgresDatabaseMonitoringRole8AB0DCC9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "monitoring.rds.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonRDSEnhancedMonitoringRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Production",
          },
          {
            "Key": "Project",
            "Value": "test-repo",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-repo-testrds-production-stack",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "apiParameterPostgresConnection9FF9728D": {
      "Properties": {
        "AllowedPattern": ".*",
        "Description": "PostgreSQL string connection TestRepo DB Production",
        "Name": "/M47.TestRepo.Apps.Minimal.Api/Production/Database/ConnectionStrings/TestdbPostgres",
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Fn::Join": [
            "",
            [
              "host=",
              {
                "Fn::GetAtt": [
                  "PostgresDatabase0A8A7373",
                  "Endpoint.Address",
                ],
              },
              ";user=testuser;dbname=testdb;port=",
              {
                "Fn::GetAtt": [
                  "PostgresDatabase0A8A7373",
                  "Endpoint.Port",
                ],
              },
              ";password={{resolve:secretsmanager:",
              {
                "Ref": "postgresqlCredentials71085D2E",
              },
              ":SecretString:password::}};sslmode=require;",
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "databaseSecurityGroupC7B2A60C": {
      "Properties": {
        "GroupDescription": "test-repoDatabase",
        "GroupName": "test-repoDatabase",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": {
              "Fn::ImportValue": "TestStack:ExportsOutputFnGetAttMockVpc92466B7FCidrBlock38F9A6B9",
            },
            "Description": "Allow PostgreSQL inbound from VPC",
            "FromPort": 5432,
            "IpProtocol": "tcp",
            "ToPort": 5432,
          },
          {
            "CidrIp": "10.0.0.0/16",
            "Description": "Allow PostgreSQL inbound from legacy VPC",
            "FromPort": 5432,
            "IpProtocol": "tcp",
            "ToPort": 5432,
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "TestStack:ExportsOutputRefMockVpc92466B7FEBC0349C",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "databaseSecurityGroupfromTestStackm47testrepotestrdsproductionstackdatabaseSecurityGroup2B25C36E54325432C08C8853": {
      "Properties": {
        "Description": "tcp5432 PostgreSQL",
        "FromPort": 5432,
        "GroupId": {
          "Fn::GetAtt": [
            "databaseSecurityGroupC7B2A60C",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "databaseSecurityGroupC7B2A60C",
            "GroupId",
          ],
        },
        "ToPort": 5432,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "postgresqlCredentials71085D2E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Description": "testdbPostgreSQL Database Credentials",
        "GenerateSecretString": {
          "ExcludeCharacters": ""@/\\ ';:,=*!()$",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{"username":"testuser"}",
        },
        "Name": "testdbPostgresqlCredentialsProduction",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "postgresqlCredentialsAttachment0038AC2B": {
      "Properties": {
        "SecretId": {
          "Ref": "postgresqlCredentials71085D2E",
        },
        "TargetId": {
          "Ref": "PostgresDatabase0A8A7373",
        },
        "TargetType": "AWS::RDS::DBInstance",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "workerParameterPostgresConnection552F442F": {
      "Properties": {
        "AllowedPattern": ".*",
        "Description": "PostgreSQL string connection TestRepo DB Production",
        "Name": "/M47.TestRepo.Apps.WorkerService/Production/Database/ConnectionStrings/TestdbPostgres",
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Fn::Join": [
            "",
            [
              "host=",
              {
                "Fn::GetAtt": [
                  "PostgresDatabase0A8A7373",
                  "Endpoint.Address",
                ],
              },
              ";user=testuser;dbname=testdb;port=",
              {
                "Fn::GetAtt": [
                  "PostgresDatabase0A8A7373",
                  "Endpoint.Port",
                ],
              },
              ";password={{resolve:secretsmanager:",
              {
                "Ref": "postgresqlCredentials71085D2E",
              },
              ":SecretString:password::}};sslmode=require;",
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
